File: Python/BOJ/재귀/1992.py
import sys
readl = sys.stdin.readline

def DaC(x, y, size):
  global result
  
  item = board[x][y]
  if size == 1:
    result += str(item)
    return
  
  isDevide = False
  for i in range(x, x + size):
    if isDevide: break
      
    for j in range(y, y + size):
      if board[i][j] != item:
        isDevide = True
        break

  if isDevide:
    result += "("
    
    half = int(size / 2)
    DaC(x, y, half)
    DaC(x, y + half, half)
    DaC(x + half, y, half)
    DaC(x + half, y + half, half)
    
    result += ")"
  else:
    result += str(item)
  
n = int(readl())
board = [list(map(int, readl().rstrip())) for _ in range(n)]

result = ""
DaC(0, 0, n)
print(result)

File: Swift/BOJ/1000/1992.swift
//
//  1992.swift
//  SwiftAlgorithm
//
//  Created by 유정주 on 2023/08/08.
//

import Foundation

func quard(_ x: Int, _ y: Int, _ size: Int) -> String {
    let half = size / 2
    
    var item = arr[x][y]
    var str = ""
    for i in x..<x+size {
        for j in y..<y+size {
            if arr[i][j] != item {
                str += quard(x, y, half)
                str += quard(x, y + half, half)
                str += quard(x + half, y, half)
                str += quard(x + half, y + half, half)
                return "(\(str))"
            }
        }
    }
    
    return "\(item)"
}

let n = Int(readLine()!)!
var arr: [[Int]] = []
for _ in 0..<n {
    let item = readLine()!.map { Int(String($0))! }
    arr.append(item)
}

let result = quard(0, 0, n)
print(result)
